File: scripts/getparams.R
#    TimeAdapt: joint inference of demography and selection
#    Copyright (C) 2021  Miguel de Navascués, Uppsala universitet, INRAE
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

library(extraDistr, quietly=TRUE)
library(ini, quietly = TRUE)
source("scripts/timeadapt.R")

# read command line arguments
args = commandArgs(trailingOnly=TRUE)
if (length(args)==0){
  stop("Two positional arguments necessary in the command (RDS options file and batch number)")
  quit(save="no")
} else {
  options_file = args[1]             # options_file = "results/test/project_options.RDS"
  batch        = as.integer(args[2]) # batch = 1
}

opts = readRDS(options_file)
Settings = opts$Settings
Model    = opts$Model
Priors   = opts$Priors
Sample   = opts$Sample
Genome   = opts$Genome

total_periods = Model$periods_forward+Model$periods_coalescence

# print script info to screen
print_info("getparams.R", Settings$verbose, project = Settings$project, batch = batch)

# set seed for random number generator and get seeds for msprime and SLiM
set.seed(Settings$seed + as.numeric(batch))
seed_coal = as.integer(round(runif(Settings$num_of_sims, 0, 2^31)))
seed_forw = as.integer(round(runif(Settings$num_of_sims, 0, 2^31)))
seed_mut  = as.integer(round(runif(Settings$num_of_sims, 0, 2^31)))
if (Settings$verbose >= 10) write(paste("Seed (msprime-coalescence)", seq_len(Settings$num_of_sims), ":", seed_coal), stdout())
if (Settings$verbose >= 10) write(paste("Seed (SLiM)", seq_len(Settings$num_of_sims), ":", seed_forw), stdout())
if (Settings$verbose >= 10) write(paste("Seed (msprime-mutation)", seq_len(Settings$num_of_sims), ":", seed_mut), stdout())

# create results directory
batch_dir <- paste(Settings$project_dir, batch, sep="/")
dir.create(batch_dir, showWarnings = FALSE)

#### SAMPLE FROM PRIORS #### 

# Generation length = generation interval in years
gen_len = rnsbeta(Settings$num_of_sims,
                  Priors$gen_len_sh1,
                  Priors$gen_len_sh2,
                  Priors$gen_len_min,
                  Priors$gen_len_max)
if (Settings$verbose >= 10) write(paste("Generation length", seq_len(Settings$num_of_sims), ":", gen_len), stdout())
# mutation rate
u = rlnorm(Settings$num_of_sims,
           log(Priors$mut_rate_mean),
           Priors$mut_rate_sd)
if (Settings$verbose >= 10) write(paste("Mutation rate", seq_len(Settings$num_of_sims), ":", u), stdout())

# census population size (empty data frame)
N = as.data.frame(matrix(nrow = Settings$num_of_sims,
                         ncol = total_periods))
names(N) = paste0("N_",seq_len(total_periods))

# ages of samples in generations (empty data frame)
ages = as.data.frame(matrix(nrow = Settings$num_of_sims,
                            ncol = Sample$size))
names(ages) = paste0(Sample$id,"_age")

for (sim in seq_len(Settings$num_of_sims)){
  if (Settings$verbose >= 10) write(paste("Simulation", sim), stdout())
  
  # sample demographic (N) trajectory
  N_trajectory = round(sample_param_trajectory(total_periods,
                                               Priors$pop_size_min,
                                               Priors$pop_size_max))
  N[sim,] = N_trajectory
  backward_demography = rev(N_trajectory[1:Model$periods_coalescence])
  forward_demography = N_trajectory[(Model$periods_coalescence+1):total_periods]
  if (Settings$verbose >= 10){
    write("Demography (N):", stdout())  
    write(paste(N_trajectory, collapse = " "), stdout())  
    write("Demography on coalscent period:", stdout())  
    write(paste(backward_demography, collapse = " "), stdout())  
  }
  
  # get ages of samples in number of generation in the past
  # ancient sample ages are taken from the calibrated age PDF
  sample_ages = get_sample_ages(Sample, gen_len[sim])
  ages[sim,] = sample_ages
  chrono_order = order(sample_ages)-1
  sampling_times_msprime = sort(unique(sample_ages))
  sampling_times_slim = rev(Model$generations_forward - sampling_times_msprime)
  sample_size_per_gen = rep(NA,length(sampling_times_msprime))
  for (i in seq_along(sampling_times_msprime)){
    sample_size_per_gen[i] = sum(sample_ages==sampling_times_msprime[i])
  }
  if (Settings$verbose >= 10){
    write("Ages of samples (in generations):", stdout())  
    write(paste(sample_ages, collapse = " "), stdout())  
    write("Chronological order of samples:", stdout())  
    write(paste(chrono_order, collapse = " "), stdout())  
    write("Sample size at each sampling time:", stdout())  
    write(paste(sample_size_per_gen, collapse = " "), stdout())  
  }

  # write sim *.ini file
  sim_ini <- list()
  sim_ini[["Simulation"]] = list(batch        = batch,
                                 sim          = sim)
  sim_ini[["Sample"]]     = list(ss           = paste(sample_size_per_gen, collapse=" "),
                                 msprime_ts   = paste(sampling_times_msprime, collapse=" "),
                                 chrono_order = paste(chrono_order, collapse=" ")) 
  sim_ini[["Demography"]] = list(N            = paste(backward_demography, collapse=" ")) 
  sim_ini[["Genome"]]     = list(mu           = u[sim],
                                 ttratio      = 2.0,
                                 seq_error    = 0.005)
  sim_ini[["Seeds"]]      = list(seed_coal    = seed_coal[sim],
                                 seed_mut     = seed_mut[sim])
  
  sim_ini_file <- paste0(batch_dir,"/sim_",sim,".ini")
  write.ini(sim_ini, sim_ini_file)
  
  # write source file for SLiM
  source4slim <- paste0("setSeed(", seed_forw[sim], ");\n",
                        "defineConstant(\"L\",", Genome$L, ");\n",
                        "defineConstant(\"N\",c(", paste(forward_demography, collapse = ","), "));\n",
                        "defineConstant(\"tc\",c(", paste(Model$times_of_change_forw, collapse = ","), "));\n",
                        "defineConstant(\"ts\",c(", paste(sampling_times_slim, collapse = ","), "));\n",
                        "defineConstant(\"ss\",c(", paste(rev(sample_size_per_gen), collapse = ","), "));\n",
                        "defineConstant(\"np\",", Model$periods_forward, ");\n",
                        "defineConstant(\"na\",", length(sampling_times_slim) - 1, ");\n",
                        "defineConstant(\"i\",", sim, ");\n",
                        "defineConstant(\"batch\",", batch, ");\n",
                        "defineConstant(\"project\",\"", Settings$project, "\");\n",
                        "defineConstant(\"ends\",c(", paste(Genome$slim_r_map$positions, collapse = ","), "));\n",
                        "defineConstant(\"rates\",c(", paste(Genome$slim_r_map$rates, collapse = ","), "));\n")
  write(source4slim, file = paste0(batch_dir, "/sim_", sim, ".eidos"))
}

params_RDS_file = paste0(batch_dir,"/ref_table_params.RDS")
params = cbind(N,u,gen_len,ages)
saveRDS(params,file=params_RDS_file)

File: scripts/getparams.py
#    TimeAdapt: joint inference of demography and selection
#    Copyright (C) 2021  Miguel de Navascués, Uppsala universitet, INRAE
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
import sys
import os
import configparser  # for writing ini files
import numpy as np
import pandas as pd
import dill
import scipy.stats as st
from datetime import date
import timeadapt

def main():
  # get options for project and simulation:
  options = timeadapt.get_project_options(proj_options_file = sys.argv[1])
  sims  = range(0,options["num_of_sims"])
  batch = sys.argv[2]
  
  # print program name
  timeadapt.print_info(sys.argv[0],options["verbose"],batch=batch)

  # set seed for RNG
  np.random.seed(options["seed"])
  
  # create results directory and project/batch subdirectories
  try : os.mkdir("results")
  except FileExistsError : 
    if options["verbose"] >=1 : print("results folder already exists")
  project_dir = "results/"+options["project"]  
  try : os.mkdir(project_dir)
  except FileExistsError : 
    if options["verbose"] >=1 : print("{} folder already exists".format(project_dir))
  batch_dir = project_dir+"/"+str(batch)
  try : os.mkdir(batch_dir)
  except FileExistsError : 
    if options["verbose"] >=1 : print("{} folder already exists".format(batch_dir))

  # read sample and genome information files
  sample = timeadapt.read_sample_info(sample_info_file=options["sample_file"])
  genome_info = timeadapt.get_genome_info(options["genome_file"])

  # create Eidos file with SLiM recombination map
  with open(project_dir+'/recombination_map.eidos', 'w') as rec_map_file:
    rec_map_file.write('defineConstant("ends",c(' + str(','.join(map(str, genome_info["slim_r_map"]["positions"] ))) + '));\n')
    rec_map_file.write('defineConstant("rates",c(' + str(','.join(map(str, genome_info["slim_r_map"]["rates"] ))) + '));\n')

  # number of generations to simulate in forward (in SLiM)
  if options["verbose"] >=10 : print("generations_forward:"+str(options["generations_forward"]))
  if options["verbose"] >=10 : print("times_of_change_forw:"+str(options["times_of_change_forw"]))

  # calculate probability distribution curves for calibrated age of ancient samples
  # (from 14C ages)
  if any(sample["is_ancient"]):
    age14C_NAremoved = [x for x in sample["age14C"] if np.isnan(x) == False]
    age14Cerror_NAremoved = [x for x in sample["age14Cerror"] if np.isnan(x) == False]
    age_pdf = timeadapt.get_age_pdf(age14C_NAremoved,age14Cerror_NAremoved,'shcal13')
    if options["verbose"] >=100 : print(age_pdf)
  else:
    age_pdf=None
    
  # get the oldest possible age of samples and verify that falls within forward time simulation period
  if any(sample["is_modern"]):
    oldest_sample_age = [min(sample["ageBCAD"])]
    if any(sample["is_ancient"]):
      for i in range(0,sum(sample["is_ancient"])):
        if options["verbose"] >=10 : print("oldest_sample_age " + str(oldest_sample_age))
        if options["verbose"] >=10 : print("age_pdf " + str(age_pdf))
        oldest_sample_age = [min(oldest_sample_age + age_pdf[i]["ageBCAD"])]
  else:
    oldest_sample_age = [date.today().year]
    for i in range(0,sum(sample["is_ancient"])):
      oldest_sample_age = [min(oldest_sample_age + age_pdf[i]["ageBCAD"])]
  oldest_sample_age = abs(oldest_sample_age-sample["t0"])/options["gen_len_min"]
  assert oldest_sample_age+2 < options["generations_forward"],"Larger period in forward simulation necessary to include oldest sample"
  if options["verbose"] >=10 : print("oldest_sample_age " + str(oldest_sample_age))
  if options["verbose"] >=10 : print("generations_forward " + str(options["generations_forward"]))

  total_number_of_periods = options["periods_forward"] + options["periods_coalescence"]

  # create latent variable file with headers
  latent_variables_file = batch_dir+"/latent_variables.txt"
  header = "sim"
  for i in range(0,options["periods_forward"]):
    header = header + " Ne_" + str(i)
  header = header+'\n'
  with open(batch_dir+"/latent_variables.txt", 'w') as latent_variables_file:
    latent_variables_file.write(header)

  ###### SAMPLE FROM PRIORS ############################
  # sample generation length (generation time)
  gen_len = st.beta.rvs(a=options["gen_len_sh1"],
                        b=options["gen_len_sh2"],
                        loc=options["gen_len_min"],
                        scale=options["gen_len_max"]-options["gen_len_min"],
                        size=options["num_of_sims"])
  if options["verbose"] >=10 : print("gen_len " + str(gen_len))
  # sample mutation rate
  u = st.lognorm.rvs(s=options["mut_rate_sd"],
                     scale=options["mut_rate_mean"],
                     size=options["num_of_sims"])
  # sample ages and demography for each simulation
  N = [None]*options["num_of_sims"]
  for sim in sims:
    if options["verbose"] >=0 : print("simulation "+str(sim+1))
    # sample seeds for msprime and SLiM
    seed_coal = np.random.randint(1, 2**32-1)
    seed_forw = np.random.randint(1, 2**32-1)
    seed_mut  = np.random.randint(1, 2**32-1)
    if options["verbose"] >=10 : print("seeds: "+str(seed_coal)+" -- "+str(seed_forw)+" -- "+str(seed_mut))
    # sample ages in generations
    sample_ages = timeadapt.get_sample_ages(sample,age_pdf,gen_len[sim])
    sampling_times_slim = sorted([options["generations_forward"]-age for age in set(sample_ages)])
    sampling_times_msprime = sorted(set(sample_ages))
    if options["verbose"] >=10 : print("sample_ages: "+str(sample_ages))
    # get chronological order of samples and sample size per generation
    chrono_order = sorted(range(len(sample_ages)), key=lambda k: sample_ages[k])
    if options["verbose"] >=10 : print("chrono_order: "+str(chrono_order))
    sample_size_per_gen = [sample_ages.count(age) for age in sorted(set(sample_ages))]
    if options["verbose"] >=10 : print("sample_size_per_gen: "+str(sample_size_per_gen))
    # sample effective population size
    N[sim] = timeadapt.sample_param_trajectory(times=total_number_of_periods,
                                               minimum=options["pop_size_min"],
                                               maximum=options["pop_size_max"]).astype(int)
    if options["verbose"] >=10 : print("N at present simulation: "+str(N[sim]))
    
    backward_demography=np.flip(N[sim][0:options["periods_coalescence"]])
    
    sim_config = configparser.ConfigParser()
    sim_config['Simulation'] = {'sim'          : sim+1,
                                'batch'        : batch}
    sim_config['Sample']     = {'ss'           : ' '.join(map(str, sample_size_per_gen)),
                                'msprime_ts'   : ' '.join(map(str, sampling_times_msprime)),
                                'chrono_order' : ' '.join(map(str, chrono_order))}
    sim_config['Demography'] = {'N'            : ' '.join(map(str, backward_demography))}
    sim_config['Genome']     = {'mu'           : u[sim],
                                'ttratio'      : 2.0,
                                'seq_error'    : 0.005}
    sim_config['Seeds']      = {'seed_coal'    : seed_coal,
                                'seed_mut'     : seed_mut}
    with open(batch_dir+'/sim_'+str(sim+1)+'.ini', 'w') as msprime_config_file:
      sim_config.write(msprime_config_file)

    forward_demography = N[sim][total_number_of_periods-options["periods_forward"]:total_number_of_periods]

    with open(batch_dir+'/sim_'+str(sim+1)+'.eidos', 'w') as slim_config_file:
      slim_config_file.write('setSeed(' + str(seed_forw) + ');\n')
      slim_config_file.write('defineConstant("L",' + str(genome_info["L"]) + ');\n')
      slim_config_file.write('defineConstant("N",c(' + str(','.join(map(str, forward_demography)) ) + '));\n')
      slim_config_file.write('defineConstant("tc",c(' + str(','.join(map(str, options["times_of_change_forw"])) ) + '));\n')
      slim_config_file.write('defineConstant("ts",c(' + str(','.join(map(str, sampling_times_slim ))) + '));\n')
      slim_config_file.write('defineConstant("ss",c(' + str(','.join(map(str, reversed(sample_size_per_gen) ))) + '));\n')
      slim_config_file.write('defineConstant("np",' + str(options["periods_forward"]) + ');\n')
      slim_config_file.write('defineConstant("na",' + str(len(sampling_times_slim)-1) + ');\n')
      slim_config_file.write('defineConstant("i",' + str(sim+1) + ');\n')
      slim_config_file.write('defineConstant("batch",' + str(batch) + ');\n')
      slim_config_file.write('defineConstant("project","' + str(options["project"]) + '");\n')
      slim_config_file.write('source("'+project_dir+'/recombination_map.eidos");\n')

  # export reference table: parameters
  ref_table_params = pd.DataFrame(data    = N,
                                  index   = range(1,options["num_of_sims"]+1),
                                  columns = ["N"+str(x) for x in range(0,total_number_of_periods)])
  ref_table_params = ref_table_params.assign(u=u)
  ref_table_params = ref_table_params.assign(generation_length=gen_len)
  if options["verbose"] >=0 : print(ref_table_params)
  dill.dump(ref_table_params, file = open("results/" + options["project"] + "/" + str(batch) + "/ref_table_params.pkl", "wb"))

############################################################################################################
if __name__ == "__main__":
    main()

